server/service/campaigns_test.go

```diff
@@ -0,0 +1,197 @@
+	}
+	_ = query2ObsCannotRun
+
+	ds.NewQueryFunc = func(ctx context.Context, query *fleet.Query, opts ...fleet.OptionalArg) (*fleet.Query, error) {
+		return query, nil
+	}
+	ds.AppConfigFunc = func(ctx context.Context) (*fleet.AppConfig, error) {
+		return &fleet.AppConfig{ServerSettings: fleet.ServerSettings{LiveQueryDisabled: false}}, nil
+	}
+	ds.NewDistributedQueryCampaignFunc = func(ctx context.Context, camp *fleet.DistributedQueryCampaign) (*fleet.DistributedQueryCampaign, error) {
+		return camp, nil
+	}
+	ds.NewDistributedQueryCampaignTargetFunc = func(ctx context.Context, target *fleet.DistributedQueryCampaignTarget) (*fleet.DistributedQueryCampaignTarget, error) {
+		return target, nil
+	}
+	ds.HostIDsInTargetsFunc = func(ctx context.Context, filters fleet.TeamFilter, targets fleet.HostTargets) ([]uint, error) {
+		return nil, nil
+	}
+	ds.HostIDsByNameFunc = func(ctx context.Context, filter fleet.TeamFilter, names []string) ([]uint, error) {
+		return nil, nil
+	}
+	ds.LabelIDsByNameFunc = func(ctx context.Context, names []string) ([]uint, error) {
+		return nil, nil
+	}
+	ds.CountHostsInTargetsFunc = func(ctx context.Context, filters fleet.TeamFilter, targets fleet.HostTargets, now time.Time) (fleet.TargetMetrics, error) {
+		return fleet.TargetMetrics{}, nil
+	}
+	ds.NewActivityFunc = func(ctx context.Context, user *fleet.User, activityType string, details *map[string]interface{}) error {
+		return nil
+	}
+	ds.QueryFunc = func(ctx context.Context, id uint) (*fleet.Query, error) {
+		if id == 1 {
+			return query1ObsCanRun, nil
+		}
+		if id == 2 {
+			return query2ObsCannotRun, nil
+		}
+		return &fleet.Query{ID: 8888, AuthorID: ptr.Uint(6666)}, nil
+	}
+
+	testCases := []struct {
+		name                   string
+		user                   *fleet.User
+		teamID                 *uint // to use as host target
+		shouldFailRunNew       bool
+		shouldFailRunObsCan    bool
+		shouldFailRunObsCannot bool
+	}{
+		{
+			"global admin",
+			&fleet.User{GlobalRole: ptr.String(fleet.RoleAdmin)},
+			nil,
+			false,
+			false,
+			false,
+		},
+		{
+			"global maintainer",
+			&fleet.User{GlobalRole: ptr.String(fleet.RoleMaintainer)},
+			nil,
+			false,
+			false,
+			false,
+		},
+		{
+			"global observer",
+			&fleet.User{GlobalRole: ptr.String(fleet.RoleObserver)},
+			nil,
+			true,
+			false,
+			true,
+		},
+		{
+			"team maintainer",
+			teamMaintainer,
+			nil,
+			false,
+			false,
+			false,
+		},
+		// TODO(mna): I don't think this is properly validated by rego?
+		/*
+			{
+				"team admin, target not set to own team",
+				&fleet.User{Teams: []fleet.UserTeam{{Team: fleet.Team{ID: 1}, Role: fleet.RoleAdmin}}},
+				ptr.Uint(2),
+				false,
+				false,
+				true,
+			},
+		*/
+		{
+			"team admin, target set to own team",
+			&fleet.User{Teams: []fleet.UserTeam{{Team: fleet.Team{ID: 1}, Role: fleet.RoleAdmin}}},
+			ptr.Uint(1),
+			false,
+			false,
+			false,
+		},
+		// TODO(mna): I don't think this is properly validated by rego?
```

Comment:
Regarding those 2 TODOs, the rego file looks like this (similar for team observer and team admin):

```
# Team admin and maintainer running a non-observers_can_run query must have the targets
# filtered to only teams that they maintain
allow {
  object.type == "query"
  # If role is maintainer on any team
  team_role(subject, subject.teams[_].id) == [admin,maintainer][_]
  action == run
}
```

But AFAICT in no way does it validate the targets set on the endpoint, only that the subject is admin or maintainer of a team. Not sure if this is a bug or an incorrect comment.

Reply:
Let's create an issue for it for now.

Reply:
https://github.com/fleetdm/fleet/issues/3950